src/classlibnative/bcltype/bignum.cpp

```diff
@@ -0,0 +1,597 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+// File: bignum.cpp
+//
+
+//
+
+#include "bignum.h"
+
+UINT32 BigNum::m_power10UInt32Table[UINT32POWER10NUM] =
+{
+    1,          // 10^0
+    10,         // 10^1
+    100,        // 10^2
+    1000,       // 10^3
+    10000,      // 10^4
+    100000,     // 10^5
+    1000000,    // 10^6
+    10000000,   // 10^7
+};
+
+BigNum BigNum::m_power10BigNumTable[BIGPOWER10NUM];
+BigNum::StaticInitializer BigNum::m_initializer;
+
+BigNum::BigNum()
+    :m_len(0)
+{
+}
+
+BigNum::BigNum(UINT32 value)
+{
+    SetUInt32(value);
+}
+
+BigNum::BigNum(UINT64 value)
+{
+    SetUInt64(value);
+}
+
+BigNum::~BigNum()
+{
+}
+
+BigNum& BigNum::operator=(const BigNum &rhs)
+{
+    UINT8 length = rhs.m_len;
+    UINT32* pCurrent = m_blocks;
+    const UINT32* pRhsCurrent = rhs.m_blocks;
+    const UINT32* pRhsEnd = pRhsCurrent + length;
+
+    while (pRhsCurrent != pRhsEnd)
+    {
+        *pCurrent = *pRhsCurrent;
+
+        ++pCurrent;
+        ++pRhsCurrent;
+    }
+
+    m_len = length;
+
+    return *this;
+}
+
+int BigNum::Compare(const BigNum& lhs, UINT32 value)
+{
+    if (lhs.m_len == 0)
+    {
+        return value == 0 ? 0 : -1;
+    }
+
+    UINT32 lhsValue = lhs.m_blocks[0];
+
+    if (lhsValue > value || lhs.m_len > 1)
+    {
+        return 1;
+    }
+
+    if (lhsValue < value)
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+int BigNum::Compare(const BigNum& lhs, const BigNum& rhs)
+{
+    int lenDiff = (int)lhs.m_len - (int)rhs.m_len;
+    if (lenDiff != 0)
+    {
+        return lenDiff;
+    }
+
+    if (lhs.m_len == 0 && rhs.m_len == 0)
+    {
+        return 0;
+    }
+
+    UINT8 lastIndex = lhs.m_len - 1;
+    const UINT32* pLeftStartBlock = lhs.m_blocks;
+    const UINT32* pCurrentLeftBlock = lhs.m_blocks + lastIndex;
+    const UINT32* pCurrentRightBlock = rhs.m_blocks + lastIndex;
+
+    while (true)
+    {
+        if (*pCurrentLeftBlock > *pCurrentRightBlock)
+        {
+            return 1;
+        }
+
+        if (*pCurrentLeftBlock < *pCurrentRightBlock)
+        {
+            return -1;
+        }
+
+        if (pCurrentLeftBlock == pLeftStartBlock)
+        {
+            break;
+        }
+
+        --pCurrentLeftBlock;
+        --pCurrentRightBlock;
+    }
+
+    return 0;
+}
+
+void BigNum::ShiftLeft(UINT64 input, int shift, BigNum& output)
+{
+    int shiftBlocks = shift / 32;
+    int remaningToShiftBits = shift % 32;
+
+    for (int i = 0; i < shiftBlocks; ++i)
```

Comment:
As per the `memcpy` comment. `memset` should do the right thing, regardless of size. The current max size is only 35 bytes, but I also believe this is considerably too small.